### PR TITLE
Add the SPI driver.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: firmware/src/spi.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    return;
+  }
+
+  PLIB_SPI_BufferClear(MY_SPI);
+  transfer->callback(SPI_BEGIN_TRANSFER);
+
+  transfer->state = IN_TRANSFER;
+  PLIB_SPI_FIFOInterruptModeSelect(
+      MY_SPI,
+      SPI_FIFO_INTERRUPT_WHEN_TRANSMIT_BUFFER_IS_1HALF_EMPTY_OR_MORE);
```
Comment (peternewman) [inline firmware/src/spi.c]: 1HALF?

File: tests/harmony/mocks/plib_spi_mock.h [hunk 2]
```diff
@@ -4,24 +4,56 @@
 #include <gmock/gmock.h>
 #include "peripheral/spi/plib_spi.h"
 
-class MockPeripheralSPI {
+class PeripheralSPIInterface {
```
Comment (peternewman) [inline tests/harmony/mocks/plib_spi_mock.h]: Reviewed to here.

File: firmware/src/spi.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/spi.c]: That's the harmony symbolic name, nothing we can do about it :)

**Comment by nomis52** [comment]
This is ready to go now.

File: firmware/src/spi.c (hunk 1 above)
Comment (peternewman) [inline firmware/src/spi.c]: Ah, I did wonder

File: firmware/src/spi.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * @brief SPI Event types.
+ */
+typedef enum {
+  SPI_BEGIN_TRANSFER,
+  SPI_COMPLETE_TRANSFER
+} SPIEventType;
+
+/**
+ * @brief The callback run before or after the SPI transfer begins.
```
Comment (peternewman) [inline firmware/src/spi.h]: This, and the comment above seem to conflict with the one below. Is the callback actually run pre and post, with different SPIEventTypes?

File: firmware/src/spi.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ * @brief Queue an SPI transfer.
+ * @param output The output buffer to send.
+ * @param output_length The size of the output buffer.
+ * @param input The location to store received data, may be NULL.
+ * @param input_length The length of the input data buffer.
+ * @param callback The callback run prior and post this transfer.
+ * @returns True if the transfer was scheduled, false if the queue was full.
+ *
+ * This queues a write / read SPI operation. First the data in output will be
+ * sent, then input_length worth of data will be read while 0s are sent. Both
```
Comment (peternewman) [inline firmware/src/spi.h]: If the output stage is optional, can you pass NULL in as the output buffer?

File: tests/sim/Makefile.mk [hunk 5]
```diff
@@ -4,17 +4,19 @@
 noinst_LTLIBRARIES += tests/sim/libsim.la
 
 tests_sim_libsim_la_SOURCES = tests/sim/PeripheralTimer.h \
```
Comment (peternewman) [inline tests/sim/Makefile.mk]: Can you fix the ordering of this too please?

File: tests/sim/PeripheralSPI.cpp [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    ADD_FAILURE() << "Invalid SPI " << index;
+  }
+  SPI *spi = &m_spi[index];
+  spi->enabled = false;
+  spi->has_overflowed = false;
+}
+
+bool PeripheralSPI::TransmitBufferIsFull(SPI_MODULE_ID index) {
+  if (index >= m_spi.size()) {
+    ADD_FAILURE() << "invalid SPI " << index;
```
Comment (peternewman) [inline tests/sim/PeripheralSPI.cpp]: SPaG Invalid

File: tests/sim/PeripheralSPI.cpp [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+void PeripheralSPI::MasterEnable(SPI_MODULE_ID index) {
+  if (index >= m_spi.size()) {
+    ADD_FAILURE() << "Invalid SPI " << index;
+  }
+}
+
+void PeripheralSPI::FIFOInterruptModeSelect(SPI_MODULE_ID index,
+                                            SPI_FIFO_INTERRUPT mode) {
+  if (index >= m_spi.size()) {
+    ADD_FAILURE() << "invalid spi " << index;
```
Comment (peternewman) [inline tests/sim/PeripheralSPI.cpp]: SPaG Invalid

**Comment by peternewman** [comment]
Generally LGTM

File: tests/sim/PeripheralSPI.cpp (hunk 6 above)
Comment (peternewman) [inline tests/sim/PeripheralSPI.cpp]: Done

File: tests/sim/PeripheralSPI.cpp (hunk 7 above)
Comment (peternewman) [inline tests/sim/PeripheralSPI.cpp]: Done

File: firmware/src/spi.h (hunk 3 above)
Comment (nomis52) [inline firmware/src/spi.h]: s/or/and/

File: firmware/src/spi.h (hunk 4 above)
Comment (nomis52) [inline firmware/src/spi.h]: Yup, added a comment

File: tests/sim/Makefile.mk (hunk 5 above)
Comment (nomis52) [inline tests/sim/Makefile.mk]: Done

**Comment by peternewman** [comment]
LGTM

**Pull request merged**